File: app/src/main/java/com/mvcoding/financius/util/rx/RxBus.java
/*
 * Copyright (C) 2015 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.financius.util.rx;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import rx.subjects.PublishSubject;
import rx.subjects.SerializedSubject;
import rx.subjects.Subject;

@Singleton public class RxBus {
    private final Subject<Object, Object> bus = new SerializedSubject<>(PublishSubject.create());

    @Inject public RxBus() {
    }

    public void send(Object o) {
        bus.onNext(o);
    }

    public <T> Observable<T> observe(Class<T> cls) {
        //noinspection unchecked
        return (Observable<T>) bus.filter(cls::isInstance);
    }
}

File: core-app/src/main/kotlin/com/mvcoding/expensius/RxBus.kt
/*
 * Copyright (C) 2016 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.expensius

import rx.Observable
import rx.subjects.PublishSubject
import rx.subjects.SerializedSubject
import kotlin.reflect.KClass

class RxBus {
    private val bus = SerializedSubject(PublishSubject.create<Any>())

    fun send(o: Any) = bus.onNext(o)
    fun <T : Any> observe(cls: KClass<T>): Observable<T> = bus.ofType(cls.java)
}

inline fun <reified T : Any> RxBus.observe(): Observable<T> = observe(T::class)